Rework `Entity.setCustomPropForces` and add `Entity.setCustomPropShadowForce`

adamnejm opened the pull request:
## Changes:
- Move `ENT.PhysicsSimulate` function to `starfall_prop/init.lua`
- Rework the `Entity.setCustomPropForces` to work with the new static `ENT.PhysicsSimulate`
- Add the ability to disable motion controller *(stops calling the `ENT.PhysicsSimulate`)* by providing `0` as the `mode` argument to `Entity.setCustomPropForces`.
- Update documentation to state that `Entity.setCustomPropForces` cannot be used together with `Entity.setCustomPropShadowForce`

## Additions:
- Add `Entity.setCustomPropShadowForce` as a way to use [`PhysObj.ComputeShadowControl`](https://wiki.facepunch.com/gmod/PhysObj:ComputeShadowControl) inside the `ENT.PhysicsSimulate`
- Handle `deltatime` automatically inside the `ENT.PhysicsSimulate`
- Allow setting `teleportdistance` higher than `0`, only if the player has `entities.setPos` permission
- Limit the values that can crash the game: `dampfactor` to `[0..1]`, `secondstoarrive` to `[..0.001]` and perform `checkvector` on the angle *(no crash, but entity immediately disappears, better safe than sorry)*
- Add [`SIM`](https://wiki.facepunch.com/gmod/Enums/SIM) enum

adamnejm commented:
Converting to draft for a sec, I will add docs explaining that motion controller will be disabled if you pass `0` or `nil` to the functions respectively and maybe implement `teleportdistance` if player has `entities.setPos` permissions.

thegrb93 commented on lua/entities/starfall_prop/init.lua:
```diff
@@ -15,6 +15,18 @@ function ENT:Initialize()
+	elseif self.customShadowForce then
+		self.customShadowForce.deltatime = dt
+		physObj:ComputeShadowControl(self.customShadowForce)
+		return SIM_NOTHING
```
Should these be `return VECTOR_ZERO, VECTOR_ZERO, SIM_NOTHING`?

adamnejm replied:
Looking at how other people use this hook, it looks like it supports returning `SIM_NOTHING` as the only argument as well.  
Cannot find any errors or crashes, dedicated server doesn't complain about physics either when this is used.

thegrb93 replied:
ah ok cool

adamnejm commented:
Looks like `teleportdistance` isn't problematic at all, I made it require `entities.setPos` with values higher than `0`.
Also fixed `SF.Throw` stack levels and updated docs.
Should be ready for review ;p

thegrb93 commented on lua/entities/starfall_prop/init.lua:
```diff
@@ -15,6 +15,18 @@ function ENT:Initialize()
 end
 
+function ENT:PhysicsSimulate(physObj, dt)
+	if self.customForceMode then
+		return self.customForceAngular, self.customForceLinear, self.customForceMode
+	elseif self.customShadowForce then
+		self.customShadowForce.deltatime = dt
```
Does `self.whatever` trigger __index calls? We might want to optimize against that for this function.

adamnejm replied:
Well it would only trigger it maximum of two times here on `self.customForceMode` and `self.customShadowForce`. I guess I could set `self.customPhysicsMode` to a number and limit the maximum possible `__index` lookup to one.  
  
Setting `self.customPhysicsMode` to `0` on entity initialization *(viable, because it only applies to `starfall_processor` and `starfall_prop`)* could possibly elimate the `__index` lookup altogether, however now that `StopMotionController` is used, this function simply stops being executed, the `else` statement below is just for safety in case it doesn't early enough.

adamnejm replied:
Sorry kinda sleepy, updated my comment now.

adamnejm commented:
How about this?  
I added `self.customPhysicsMode` which can be `1` for the old custom prop forces or `2` for shadow control physics. This will avoid doing `__index` lookup if shadow control physics is used.  
The numbers are not quite verbose, but it's simple enough to not warrant using enums or strings there.  
  
I also moved the `MotionController` logic to `ENT.EnableCustomPhysics` to avoid duplication and took a better care of clearing old data.

thegrb93 commented:
Can you make `function ENT:PhysicsSimulate()` branchless by setting it in the setCustomPropForces and setCustomPropShadowForce functions like before? `function ent:PhysicsSimulate()`

thegrb93 commented:
Or does that not work?

adamnejm commented:
Nah, it should work just fine.

thegrb93 commented:
Actually nevermind, I forgot that JIT probably won't like and those functions would get updated a lot probably.
